src/core/lib/security/credentials/tls/grpc_tls_certificate_provider.cc

```diff
@@ -364,6 +365,56 @@ FileWatcherCertificateProvider::ReadIdentityKeyCertPairFromFiles(
   return absl::nullopt;
 }
 
+absl::Status PrivateKeyPublicKeyMatches(const std::string& private_key_string,
+                                        const std::string& cert_string) {
+  if (private_key_string.empty()) {  // test - done
+    return absl::InvalidArgumentError("Private key string is empty!");
+  }
+  if (cert_string.empty()) {  // test - done
+    return absl::InvalidArgumentError("Certificate string is empty!");
+  }
+  const char* cert_buffer = cert_string.c_str();
+  BIO* cert_string_bio = BIO_new_mem_buf(cert_buffer, strlen(cert_buffer));
+  if (cert_string_bio == nullptr) {
+    return absl::InvalidArgumentError(  // test - no idea how to
+        "Conversion from certificate string to BIO failed!");
+  }
+  X509* x509 = PEM_read_bio_X509(cert_string_bio, nullptr, nullptr, nullptr);
+  BIO_free(cert_string_bio);
+  if (x509 == nullptr) {
+    return absl::InvalidArgumentError(  // test - done
+        "Conversion from PEM string to X509 failed!");
+  }
+  EVP_PKEY* public_key = X509_get_pubkey(x509);
+  X509_free(x509);
+  if (public_key == nullptr) {
+    return absl::InvalidArgumentError(  // test - no idea how to
+        "Extraction of public key from x.509 certificate failed!");
+  }
+  const char* private_key_buffer = private_key_string.c_str();
+  BIO* private_string_bio =
+      BIO_new_mem_buf(private_key_buffer, strlen(private_key_buffer));
+  if (private_string_bio == nullptr) {
+    EVP_PKEY_free(public_key);
+    return absl::InvalidArgumentError(  // test - no idea how to
+        "Conversion from private key string to BIO failed!");
+  }
+  EVP_PKEY* private_key =
+      PEM_read_bio_PrivateKey(private_string_bio, nullptr, nullptr, nullptr);
+  BIO_free(private_string_bio);
+  if (private_key == nullptr) {
+    EVP_PKEY_free(public_key);
+    return absl::InvalidArgumentError(  // test - done
+        "Conversion from PEM string to EVP_PKEY failed!");
+  }
+  bool result = EVP_PKEY_cmp(private_key, public_key) == 1;
+  EVP_PKEY_free(private_key);
+  EVP_PKEY_free(public_key);
+  return result ? absl::OkStatus()  // test - done
+                : absl::InvalidArgumentError(
+                      "Invalid credentials pair!");  // test - done
```

Comment:
The common practice is to use a period. So suggest replacing exclamation marks in error messages with a period.

Reply:
Noted, thanks!